scripts/SREP/describe-nodes/script.sh

```diff
@@ -0,0 +1,186 @@
+      fi
+      # Check for illegal characters in the list of nodes
+      badChars="$(echo "${ARG_ARRAY[1]}" | tr -d '[:alnum:]' | tr -d '-' | tr -d '.' | tr -d ',')"
+      if [ -n "${badChars}" ]
+      then  
+        error-out "The '--nodes' argument can only contain '.|-|,|<alphaNumeric>' characters.  These characters '${badChars}' are illegal"
+      fi
+      nodes="${ARG_ARRAY[1]}"
+      ARG_ARRAY=("${ARG_ARRAY[@]:2}")  # shift past argument and value
+      ;;
+    -l|--selector)
+      if [ -n "${selector}" ] # retest
+      then
+        error-out "Only 1 '--selector' argument can be provided"
+      fi
+      # Check for unset or empty arguments and values that look like a flag/switch
+      if [ ${#ARG_ARRAY[@]} -le 1 ] \
+        || [ -z "${ARG_ARRAY[1]}" ] \
+        || [[ "${ARG_ARRAY[1]}" =~ ^-.* ]]
+      then
+        error-out "The '--selector' argument must contain a node selector"
+      fi
+      badChars="$(echo "${ARG_ARRAY[1]}" | tr -d '[:alnum:]' | tr -d '-' | tr -d '.'  | tr -d '/'  \
+          | tr -d ',' | tr -d ':' | tr -d ' ' | tr -d '(' | tr -d ')' | tr -d '!' | tr -d '=')"
+      if [ -n "${badChars}" ] # retest
+      then
+        error-out "The '--selector' argument must not contain illegal characters.  These characters '${badChars}' are illegal"
+      fi
+      selector="${ARG_ARRAY[1]}"
+      ARG_ARRAY=("${ARG_ARRAY[@]:2}")  # shift past argument and value
+      ;;
+    -*|--*)
+      error-out "Unknown option '${ARG_ARRAY[0]}'"
+      ;;
+    *)
+      error-out "Unknown argument"
+      ;;
+  esac
+done
+
+# If --nodes is supplied use that and ignore all other arguments
+# else if --selector is supplied use that ignore the others arguments
+# else if --all is supplied use that ignore the others arguments
+# else if any of the --master or --infra or --worker arguments are supplied use them all
+if [ -z "${nodes}" ]
+then
+  if [ -n "${selector}" ] # retest
+  then
+    nodes="$(list-selector "${selector}")"
+  elif [ "${all}" = true ]
+  then
+    add2nodes "$(list-selector "node-role.kubernetes.io/master")"
+    add2nodes "$(list-selector "node-role.kubernetes.io=infra")"
+    add2nodes "$(list-selector "node-role.kubernetes.io!=infra,node-role.kubernetes.io/worker")"
+  else
+    if [ "${master}" = true ]
+    then
+      add2nodes "$(list-selector "node-role.kubernetes.io/master")"
+    fi
+    if [ "${infra}" = true ]
+    then
+      add2nodes "$(list-selector "node-role.kubernetes.io=infra")"
+    fi
+    if [ "${worker}" = true ]
+    then
+      add2nodes "$(list-selector "node-role.kubernetes.io!=infra,node-role.kubernetes.io/worker")"
+    fi
+  fi
+fi
+
+# If the nodes variable is empty, it means the no arguments were supplied on the command line
+# Or the selector(s) did not pick any nodes.  This is an error
+if [ -z "${nodes}" ]  # retest
+then
+  error-out "No nodes selected"
+fi
+
+nodes=$(echo "${nodes}" | tr "," " ")
+oc describe nodes "${nodes}"
```

Comment:
Unfortunately surrounding the `${nodes}` with quotes to satisfy shellcheck breaks the script because it treats the space-separated list of nodes as a single argument.

Instead, to do this in a way that makes both the `oc` and `shellcheck` happy, we should turn the `nodes` string into a bash array:

```suggestion
IFS=',' read -r -a nodearray <<< "${nodes}"
oc describe nodes "${nodearray[@]}"
```

(I tested this and it seemed to work)

Reply:
The issue highlighted by SpellChecker is that when the line `oc describe nodes ${nodes}` is unquoted it is prone to abuse by globbing and therefore open to malicious intent from the command line.  (Well not really because I validate the input of the command line) So the recommended approach is to quote the variable

The main issue caused by quoting `oc describe nodes "${nodes}"` the variable is that now the OC CLI only sees 1 argument.  I think the following suggestion should be enough?

Checking exceptions in https://www.shellcheck.net/wiki/SC2086

This may be a better alternative?
```suggestion
nodes=($(echo "${nodes}" | tr "," " "))
oc describe nodes "${nodes[@]}"
```

Similar to this command line POC
```
dsquirre@dsquirre-mac describe-nodes % args="-l,-a,-r,-t"
dsquirre@dsquirre-mac describe-nodes % args=($(echo "${args}" | tr "," " "))
dsquirre@dsquirre-mac describe-nodes % echo ls "${args[@]}"
ls -l -a -r -t
dsquirre@dsquirre-mac describe-nodes % ls "${args[@]}"
total 40
drwxr-xr-x  14 dsquirre  staff   448  3 Mar 13:44 ..
drwxr-xr-x  13 dsquirre  staff   416  7 Mar 15:21 test
-rw-r--r--   1 dsquirre  staff  1007 10 Mar 10:21 metadata.yaml
-rw-r--r--   1 dsquirre  staff  1925 10 Mar 10:23 README.md
drwxr-xr-x   7 dsquirre  staff   224 10 Mar 11:13 .
-rwxr--r--   1 dsquirre  staff   129 10 Mar 14:48 run.sh
-rwxr--r--   1 dsquirre  staff  6132 13 Mar 16:10 script.sh
dsquirre@dsquirre-mac describe-nodes %
```

Reply:
That suggested solution, whilst working, seems to trigger SC2207 when I run shellcheck over it, so it may be trading one style issue for another. The suggested solution I provided in my comment has no shellcheck issues in case you want to do that instead.